### ISSUE TITLE
[flutter_tools][web] PWA ServiceWorker breaks on GitHub Pages (rootPath rewrite issue with url-params?)

### ISSUE THREAD

**Issue by JELaVallee** [flutter/flutter#68449]
## Steps to Reproduce

1. Run `flutter create flutter-pwa-bug`
2. Create a GitHub repo and push the project to it
3. Create a `gh-pages` empty branch and push to GitHub repo
4. Configure the GitHub repo to use the `gh-pages` branch as the GitHub Pages root for the repo
5. Update the BASE tag's `href` attribute in `{project-root}/web/index.html` to be `/{gh-repo-name}/`
6. In the flutter-pwa-bug project, run `flutter build --release`
7. Push the contents of `{project-root}/build/web/` to the `gh-pages` branch
8. Open the GitHub Pages URL in Chrome browser

I've made a repository to demonstrate this issue here: https://github.com/JELaVallee/hello-flutter-pwa
With the deployment of such here: https://jelavallee.github.io/hello-flutter-pwa

**Expected results:** 

1. In the Chrome browser, the web site loads and the application is functional
2. In the Chrome browser's location field the "+" icon is displayed
3. Clicking the "+" icon opens the "Install app?" popup dialog and allows you to install the PWA to your computer

**Actual results:** 

1. The page loads and the application functions as expected
2. The "+" icon is _not_ displayed in the browser's location field
3. In the Chrome Developer Tools, Console pane, there is the error: "Uncaught (in promise) TypeError: Request failed ... flutter_service_worker.js?v={cache-buster}"
4. In the Chrome Developer Tools, Network pane, there is a 404 error for the request: "https://<user>.github.io/?revision={app-cache-buster}" which is thrown by the `flutter_service_worker.js` when it's loading the app-cache.

**Triage Notes**

While working to narrow down this issue, I've made the following observations:

1. Hard-coding of "/" as one of the app-cache keys here: [web.dart:Line 438](https://github.com/flutter/flutter/blob/345188d40e576ffb6c97fee60446af58b8c9965a/packages/flutter_tools/lib/src/build_system/targets/web.dart#L438)
    1. In a site deployment like GitHub Pages (w/o a CNAME binding, of course) this absolute root path isn't what we want
    1. This cannot be easily over-ridden and isn't respecting the BASE:href attribute in the `index.html`
    1. I've been able to hack this to work in the generated `flutter_service_worker.js` as seen in [this commit](https://github.com/JELaVallee/hello-flutter-pwa/commit/5f8695bdd0982b7877100098dda1bf916f3715f3) which is not currently deployed...
1. The "fetch" call in `flutter_service_worker.js` (generated in [web.dart:Line 633](https://github.com/flutter/flutter/blob/345188d40e576ffb6c97fee60446af58b8c9965a/packages/flutter_tools/lib/src/build_system/targets/web.dart#L633) is throwing errors which yields the "Uncaught (in promise) TypeError" documented above on the 404.
1. ~GitHub Pages does not rewrite the rootPath URL to "index.html" when the URL has params applied to it.~
    1. ~When tested on a GitHub Pages site with a bound CNAME record, the request to "https://{domain}/?revision={app-cache-buster}" the server still returns a 404 error response.~ (EDIT: This was a bad test on my part, GitHub Pages handles these rewrites properly...)

Suggesting that we either make "/" key in observation-1 be parsed from the BASE:href in the coded `{project-root}/web/index.html` _or_ making the cache loading/checking code in the generated `flutter_service_worker.js` a bit more defensive when handling response errors vs. just throwing them up-stack.

Having one heck of a time properly testing this in the web_test harness.  Going to look at a `flutter_service_worker.js` error handling fix first since it seems less "intrusive", but thinking that a fix for observation-1 would be ideal.

Possibly related to #50580 or #52901

<details>
  <summary>Logs</summary>

```
$ flutter build web --release --verbose
[ +125 ms] executing: [/home/jelavallee/workspace/sandbox/flutter/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +62 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] 4fa4f91d5cc65a5a98c7ccd91bb7d47814048a57
[        ] executing: [/home/jelavallee/workspace/sandbox/flutter/flutter/] git tag --points-at HEAD
[  +30 ms] Exit code 0 from: git tag --points-at HEAD
[        ] 1.23.0-13.0.pre
[  +17 ms] executing: [/home/jelavallee/workspace/sandbox/flutter/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +18 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/dev
[        ] executing: [/home/jelavallee/workspace/sandbox/flutter/flutter/] git ls-remote --get-url origin
[   +4 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[ +160 ms] executing: [/home/jelavallee/workspace/sandbox/flutter/flutter/] git rev-parse --abbrev-ref HEAD
[  +19 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] dev
[ +174 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +3 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +113 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ] Artifact Instance of 'AndroidMavenArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[   +4 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
[ +233 ms] Generating
/home/jelavallee/workspace/sandbox/hello-flutter-pwa/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[ +129 ms] Generating
/home/jelavallee/workspace/sandbox/hello-flutter-pwa/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[   +5 ms] Compiling lib/main.dart for the Web...
[   +4 ms] Initializing file store
[   +7 ms] Done initializing file store
[ +106 ms] Skipping target: gen_localizations
[  +66 ms] Skipping target: web_entrypoint
[ +569 ms] Skipping target: dart2js
[  +48 ms] web_release_bundle: Starting due to {InvalidatedReason.outputMissing}
[ +168 ms] web_release_bundle: Complete
[ +100 ms] web_service_worker: Starting due to {InvalidatedReason.inputChanged, InvalidatedReason.outputMissing}
[  +80 ms] web_service_worker: Complete
[  +10 ms] Persisting file store
[  +15 ms] Done persisting file store
[  +57 ms] Compiling lib/main.dart for the Web... (completed in 1.2s)
[   +2 ms] "flutter web" took 1,757ms.
[ +112 ms] ensureAnalyticsSent: 108ms
[  +62 ms] Running shutdown hooks
[        ] Shutdown hooks complete
[        ] exiting with code 0
```

```
 $ flutter analyze
Analyzing hello-flutter-pwa...                                          
No issues found! (ran in 2.4s)
```

```
$ flutter doctor -v
[✓] Flutter (Channel dev, 1.23.0-13.0.pre, on Linux, locale en_US.UTF-8)
    • Flutter version 1.23.0-13.0.pre at /home/jelavallee/workspace/sandbox/flutter/flutter
    • Framework revision 4fa4f91d5c (12 days ago), 2020-10-06 12:31:25 -0700
    • Engine revision 443cd5a1e1
    • Dart version 2.11.0 (build 2.11.0-186.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/jelavallee/Android/Sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /home/jelavallee/Android/Sdk
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /usr/bin/chromium

[✓] Android Studio (version 3.5)
    • Android Studio at /opt/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.50.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.15.1

[✓] Connected device (3 available)
    • nami (mobile)    • emulator-5554 • android-x64    • Android 9 (API 28)
    • Web Server (web) • web-server    • web-javascript • Flutter Tools
    • Chrome (web)     • chrome        • web-javascript • Chromium 83.0.4103.116 built on Debian 10.4, running on Debian 10.6
```

</details>

**Comment by JELaVallee** [comment]
After reading [this awesome article](https://web.dev/offline-cookbook/) and getting a little more aggressive with the Chrome Dev Tools script debugger, it's clear that this is intended behavior for the `cache.addAll(...)` method when there is a non-200 on a cached Request.

A defensive strategy on the `flutter_service_worker.js` browser install side doesn't make sense, given that.

Also, editing my original observations-3, as I was able to prove that wasn't the case... 

Going to refocus on observation-1, now.

**Comment by TahaTesser** [comment]
Hi @JELaVallee 
Opening https://jelavallee.github.io/hello-flutter-pwa in chrome, there are no errors in the console or network tab
<img width="348" alt="Screenshot 2020-10-19 at 6 07 05 PM" src="https://user-images.githubusercontent.com/48603081/96451916-e89b7180-1235-11eb-87fb-2a9d53007641.png">
Can you please reproduce same errors on this version of chrome?

**Comment by azbuky** [comment]
@TahaTesser 
Try refreshing the page with the DevTools open.

**Comment by JELaVallee** [comment]
Hey there @TahaTesser 

I'm pretty much on the same version as you (on ChromeOS and just slightly behind in my debian container):

![image](https://user-images.githubusercontent.com/4291961/96489777-e1f90400-120d-11eb-9dbf-85f6e101b9f3.png)

As @azbuky pointed out, you need to refresh after initial load to the the errors in the console (short of opening the console _first_ in a new window/tab before loading the URL)... 

Visually this is what I'm seeing:

![image](https://user-images.githubusercontent.com/4291961/96490609-115c4080-120f-11eb-82f2-b2b0f19c37e0.png)

**Comment by TahaTesser** [comment]
<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel dev, 1.24.0-1.0.pre, on Mac OS X 10.15.7 19H2 darwin-x64,
    locale en-GB)
    • Flutter version 1.24.0-1.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision 12bea61c29 (5 days ago), 2020-10-14 15:57:04 -0700
    • Engine revision 069b3cf8f0
    • Dart version 2.11.0 (build 2.11.0-218.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.10.0.rc.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 10 (API 29)
    • macOS (desktop)  • macos            • darwin-x64     • Mac OS X 10.15.7
      19H2 darwin-x64
    • Web Server (web) • web-server       • web-javascript • Flutter Tools
    • Chrome (web)     • chrome           • web-javascript • Google Chrome
      86.0.4240.80

• No issues found!

```

</details>

**Comment by kphanipavan** [comment]
![Screenshot 2020-11-03 at 7 22 44 PM](https://user-images.githubusercontent.com/60005847/97993355-f6422100-1e09-11eb-9b22-e901d6b03c43.png)
The request is supposed to be
```
https://kphanipavan.github.io/Time/?revision=01224db2a039381596e9a26bcd4c339a
```
Which verifies the hast of the root folder.
But the request which was sent was, as seen in the above image,
```
https://kphanipavan.github.io/?revision=01224db2a039381596e9a26bcd4c339a
```
I will try things which may fix it, and will report it when done.

**Comment by kphanipavan** [comment]
This might be a wrong thing to do but, 
Commenting [this line](https://github.com/kphanipavan/Time/blob/067251b8ba899c8d2fff799eac6480daa147d772/flutter_service_worker.js#L24) worked for me flawlessly...

**Comment by JELaVallee** [comment]
@kphanipavan That's confirming what I've seen as well.  Ideally, all of those "CORE" asset references would be based off of the BASE defined in the `web/index.html` working file.

In your example:
```
const CORE = [
"/Time/",
"/Time/main.dart.js",
"/Time/index.html",
"/Time/assets/NOTICES",
"/Time/assets/AssetManifest.json",
"/Time/assets/FontManifest.json"];
```

While hacking out the "/" reference _does_ get the PWA capability working, it's not something that can be easily automated or done gracefully in the build process since that "/" absolute-root reference is hardcoded in the `web.dart` generator.

**Comment by maxim-saplin** [comment]
This [PR](https://github.com/flutter/flutter/pull/52606) was supposed to fix the tooling and make both index.html and flutter_service_worker.js work fine under non-root paths (such as GH Pages). But if we look at what gets generated with the current version (I'm now on beta 2.2.0-10.1.pre), the service worker has all pathes fixed (they don't start with "/") except of two "/" entries (in RESOURCES and CORE constants):

![image](https://user-images.githubusercontent.com/7947027/116008009-3ba61500-a61b-11eb-8926-0685f06bb86e.png)

This causes service worker fetching index.html from "/" which returns 404:

![image](https://user-images.githubusercontent.com/7947027/116008083-7dcf5680-a61b-11eb-821d-171e6828a643.png)

If you manualy change the two entry to empty string (replace "/" with "") all is good:

![image](https://user-images.githubusercontent.com/7947027/116008134-aeaf8b80-a61b-11eb-8c07-265de8de3a4e.png)


Long story short, to avoid 4s delay when loading Flutter Web app from gh-pages (due to failing service worker), open this branch in GitHub UI and change the 2 entries to "". You might also need to change the Actions workflow YAML (if you happen to use GitHub actions) to avoid reruning the build process upon each commit (e.g. on: [workflow_dispatch] #[push]) and keep in mind that after each re-build the js must be fixed.

Looking forward for the fix from Flutter team.

**Comment by grant-dennison** [comment]
I ran into a problem that is related to the specific erroneous behavior observed in this issue. In my case, I was unable to convince my phone to update to the most recent version of my PWA, and it seems [I am not alone](https://stackoverflow.com/q/62018659/4639640).

Some of my issues (like `main.dart.js` never getting reloaded) may be resolved in some changes coming down the pipeline. (#75535, specifically [here](https://github.com/flutter/flutter/commit/e7953b3be496bd54cb1a00d70c1a3218b7a336b7#diff-db6d2c7a68833ee246877b3e2d05e9c567a293c294c3cf903ba69aeb6c9daaf7))

But piggy-backing off of @maxim-saplin's suggestion, I put together some further changes in #82521.

Hopefully these breadcrumbs will be useful for someone else out there. I don't have the heart right now to push through another open-source contribution process, so I don't intend to push the PR further soon.

**Comment by matsp** [comment]
For everyone who need to workaround the absolute caching e.g. CI/CD pipeline:

`sed -i 's+"/"+""+g' flutter_service_worker.js`

**Comment by steebchen** [comment]
I run into the same issue, and can confirm that the "/" path is the problem. It should probably be either "index.html" or ".", I would say. The `sed` suggestion from @matsp successfully works around this issue.

**Comment by longtimedeveloper** [comment]
I have the same issue.  The workaround here corrects the problem.  I'm new to the Flutter community and do not understand why the team has not corrected this issue after over a year.  Seems like adding a command-line option like `--deployingToGitHubPages` has not been added to the `flutter build web` command.

**Comment by steebchen** [comment]
@longtimedeveloper I feel like that's the wrong solution to the problem – the correct solution would be to consider the base bath from `/` to `.` for the first/root URL.

**Comment by longtimedeveloper** [comment]
@steebchen if you're correct, then why hasn't this been accomplished and the issue corrected and closed?

**Comment by steebchen** [comment]
I don't know, but probably because there are 

![image](https://user-images.githubusercontent.com/5013932/144645860-b13caeea-9b00-42b3-aac2-641ec105f3a7.png)
other issues, and the team can only do so much ;)

**Comment by nicolasvac** [comment]
This is probably a super easy-fix that's impacting everyone who is trying to slowly integrate flutter in already existing systems.

@yjbanov I'm sorry to ping you, but can this be scheduled earlier ?

Without this flutter always searches assets from the root path, making it impossible to use in sub directories.
This is not even documented in the wiki or something.
